src/control/destination/background/destination_set.rs

```diff
@@ -230,6 +230,10 @@ where
         );
         match self.addrs.take() {
             Exists::Yes(mut cache) => {
+                self.responders.retain(|r| {
+                    let sent = r.update_tx.unbounded_send(Update::NoEndpoints);
```

Comment:
This will send the new `Update::NoEndpoints` state that we poll for in resolve now, but why does it need to be `retain`? Shouldn't all the responders be ready to receive?

Reply:
the `retain` will drop any responders for which the receive end has been dropped. `on_change` also uses `retain`: https://github.com/linkerd/linkerd2-proxy/blob/767b8c69d0e50b8a4e14806d3bf3ede66694d6ad/src/control/destination/background/destination_set.rs#L267-L271